### ISSUE TITLE
parser: stack overflow on deeply nested statement bodies

### ISSUE THREAD

**Issue by camc314** [oxc-project/oxc#21953]
## Summary

The parser can abort with a native stack overflow when parsing extremely deep nested statement bodies. This was observed in ecosystem CI against Bun's `lots-of-for-loop.js` fixture, which contains ~15k directly nested `for (...)` statements ending in one body statement.

Failing ecosystem job: https://github.com/oxc-project/oxc-ecosystem-ci/actions/runs/25101881953/job/73554114827

Relevant Bun context:
https://github.com/oven-sh/bun/pull/15985

## Reproduction

```sh
cargo run -q -p oxc_parser --example parser -- /Users/cameron/github/oven-sh/bun/test/bundler/transpiler/fixtures/lots-of-for-loop.js
```

Current behavior:

```text
thread 'main' has overflowed its stack
fatal runtime error: stack overflow, aborting
```

A minimal generated shape:

```js
let counter = 0;
for (let i = 0; i < 1; i++) for (let i = 0; i < 1; i++) /* repeat many times */ counter++;
```

Similar risk exists for other recursively parsed statement bodies, for example:

```js
if (true) { if (true) { if (true) { /* repeat deeply */ } } }
```

and plain nested blocks.

## Expected Behavior

The parser should not abort the process. It should either:

1. Parse the AST successfully
2. Return a parser diagnostic for excessive nesting before native stack exhaustion.

## Notes

This is performance-sensitive code, so any fix should avoid regressing normal parser hot paths.

Other engine/parser behavior checked:

- V8 uses explicit stack checks in parser code and reports parser failure instead of relying on native stack overflow.
- SWC currently stack-overflows on the Bun `for` fixture. It does have selective handling for `if` consequents via `stacker::maybe_grow` and parses `else if` chains iteratively, but it does not structurally handle nested `for` bodies.
- Bun's PR added stack-overflow handling around parser/visitor paths, but Oxc should avoid relying on enlarged stack size.

Potential direction:

- Handle consecutive `for (...) for (...) ...` bodies iteratively by parsing loop headers into a temporary stack and folding the AST inside-out.
- Add a general recursive statement-depth guard for statement body recursion such as blocks, `if`, loops, labels, `with`, and switch consequents.
- Keep the common single-statement/single-loop path allocation-free.

**Comment by overlookmotel** [comment]
I just noticed this comment from Boshen: https://github.com/oxc-project/backlog/issues/58#issuecomment-2214388689

> No `stacker`, I tried it before, it doesn't work in wasm.

**Comment by overlookmotel** [comment]
I had a chat with Claude about `stacker`. It uses a thread local to store the address of the bottom of the stack. Thread locals are cheap in a normal Rust binary, but in a dynamic library (which NAPI apps are) they're quite expensive - ~30-60 cycles.

So if we *did* go with `stacker`, we'd likely need some other mechanism to track recursion depth, to avoid calling `stacker::maybe_grow` too often. Ergh.

There's a simpler solution. In our own apps (Oxlint, Oxfmt), just increase the stack size to something huge. Effect on different platforms:

* Linux: with overcommit enabled (the default) the large stack allocation will only use virtual memory, not physical memory - stack expands its committed memory only as needed.
* Mac: Ditto (it doesn't have a "no overcommit" option, so overcommit is always enabled).
* Windows: Ditto.

Notably, Claude tells me that Windows takes an overcommit approach with stack, despite its usual policy being overcommit disabled. So, if Claude is to be believed, we shouldn't run into OOM problems with this approach (i.e. no repeat of the raw transfer OOM problems due to no overcommit on WIndows).

For consumers of Oxc as libraries, we just have to tell them to increase stack size if they expect to need to handle files which contain deep recursion.

The other approach, as noted by Cam above, is to convert recursive calls into loops (Pratt parsing?). But in a recursive descent parser, that's quite a large change!
